server/block/campfire.go

```diff
@@ -0,0 +1,267 @@
+		return false
+	}
+	if _, ok := w.Liquid(pos); ok {
+		return false
+	}
+
+	c.Extinguished = false
+	w.SetBlock(pos, c, nil)
+	return true
+}
+
+// Splash ...
+func (c Campfire) Splash(w *world.World, pos cube.Pos, p potion.Potion) {
+	if p != potion.Water() || c.Extinguished {
+		// Water is the only potion that can extinguish a lit campfire.
+		return
+	}
+
+	w.PlaySound(pos.Vec3Centre(), sound.FireExtinguish{})
+	c.Extinguished = true
+	w.SetBlock(pos, c, nil)
+}
+
+// Activate ...
+func (c Campfire) Activate(pos cube.Pos, _ cube.Face, w *world.World, u item.User, ctx *item.UseContext) bool {
+	held, _ := u.HeldItems()
+	if held.Empty() {
+		return false
+	}
+	rawFood, ok := held.Item().(item.Smeltable)
+	if !ok || !rawFood.SmeltInfo().Food {
+		return false
+	}
+
+	for i, it := range c.Items {
+		if it.Item.Empty() {
+			c.Items[i] = CampfireItem{
+				Item: held.Grow(-held.Count() + 1),
+				Time: time.Second * 30,
+			}
+
+			ctx.SubtractFromCount(1)
+
+			w.PlaySound(pos.Vec3Centre(), sound.ItemAdd{})
+			w.SetBlock(pos, c, nil)
+			return true
+		}
+	}
+	return false
+}
+
+// UseOnBlock ...
+func (c Campfire) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	if _, ok := w.Block(pos).(Campfire); ok && face == cube.FaceUp {
+		return false
+	}
+	pos, _, used = firstReplaceable(w, pos, face, c)
+	if !used {
+		return
+	}
+	c.Facing = user.Rotation().Direction().Opposite()
+	place(w, pos, c, user, ctx)
+	return placed(ctx)
+}
+
+// Tick is called to cook the items within the campfire.
+func (c Campfire) Tick(_ int64, pos cube.Pos, w *world.World) {
+	if c.Extinguished {
+		// Extinguished, do nothing.
+		return
+	}
+	if rand.Float64() <= 0.016 { // Every three or so seconds.
+		w.PlaySound(pos.Vec3Centre(), sound.CampfireCrackle{})
+	}
+
+	updated := false
+	for i, it := range c.Items {
+		if it.Item.Empty() {
+			continue
+		}
+
+		updated = true
+		if it.Time > 0 {
+			c.Items[i].Time = it.Time - time.Millisecond*50
+			continue
+		}
+
+		if food, ok := it.Item.Item().(item.Smeltable); ok {
+			dropItem(w, food.SmeltInfo().Product, pos.Vec3Middle())
+		}
+		c.Items[i].Item = item.Stack{}
+	}
+	if updated {
+		w.SetBlock(pos, c, nil)
+	}
+}
+
+// NeighbourUpdateTick ...
+func (c Campfire) NeighbourUpdateTick(pos, _ cube.Pos, w *world.World) {
+	if _, ok := w.Liquid(pos); ok && !c.Extinguished {
+		c.Extinguished = true
+		w.PlaySound(pos.Vec3Centre(), sound.FireExtinguish{})
+		w.SetBlock(pos, c, nil)
+	}
+}
+
+// EntityInside ...
+func (c Campfire) EntityInside(pos cube.Pos, w *world.World, e world.Entity) {
+	if flammable, ok := e.(flammableEntity); ok {
+		if flammable.OnFireDuration() > 0 && c.Extinguished {
+			c.Extinguished = false
+			w.PlaySound(pos.Vec3(), sound.Ignite{})
+			w.SetBlock(pos, c, nil)
+		}
+		if !c.Extinguished && !w.RainingAt(pos) {
+			if flammable.OnFireDuration() < time.Second*8 {
+				flammable.SetOnFire(8 * time.Second)
+			}
+		}
```

Comment:
I don't see this behavior in vanilla or on the wiki

See https://minecraft.wiki/w/Campfire#Damage and https://minecraft.wiki/w/Soul_Campfire#Damage